### PR TITLE
Change: Improve duplicate detection for company/president name changes

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Initially I started this by wanting to remove the `goto`s within `GenerateCompanyName()` and `GeneratePresidentName()`.

However after doing so I realised that, for company names at least, the duplicate name detection appears to be inadequate as it's possible for different seed values to have the same string result. This is perhaps more theoretical than practical, however.

For duplicate company names, only the string type and seed parameters are checked.

For duplicate president names, although the actual string is compared, we unnecessarily "re-fetch" the president names for each iteration.

## Description

* Refactor both `GenerateCompanyName()` and `GeneratePresidentName()` to remove `goto`s. In particular this makes the logic of the company name generator easier to understand.
* Always test for duplicate names by comparing actual string values instead of seed parameters.
* As we always need to test for duplicate names, fetch the other names once in advance. This (probably) temporarily uses more memory but (probably) reduces CPU usage.

With the new flow it's also now clear to see that the loops are unbounded, but haven't addressed that.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/company_cmd.cpp [hunk 1]
```diff
@@ -552,29 +552,45 @@ static Colours GenerateCompanyColour()
...
+ */
+static bool SetPresidentName(std::span<const std::string> other_names, Company *c, uint32_t seed)
+{
+	assert(c != nullptr);
+
+	c->president_name_1 = SPECSTR_PRESIDENT_NAME;
+	c->president_name_2 = seed;
+
+	/* Reserve space for extra unicode character. We need to do this to be able
+	 * to detect too long president name. */
```
Comment (rubidium42) [inline src/company_cmd.cpp]: This seems to be a vestige of pre-`std::string` behaviour that should have been removed during that migration.

**Review by rubidium42 — APPROVED** [review]
